File: 687. Longest Univalue Path/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private int helper(TreeNode root, int val) {
        if (root == null) {
            return 0;
        }

        return root.val == val ? 1 : 0 + helper(root.left, val) + helper(root.right, val);
    }

    public int longestUnivaluePath(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Math.max(helper(root, root.val), Math.max(longestUnivaluePath(root.left), longestUnivaluePath(root.right)));
    }
}

File: 687. Longest Univalue Path/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int longestUnivaluePath(TreeNode* root) {
        if (!root) { return 0; }
        int m = INT_MIN;
        helper(root, root, m);
        return m;
    }
    
    int helper(TreeNode* root, TreeNode* parent, int& m) {
        if (!root) { return 0; }
        
        if (root->val == parent->val && root != parent) {
            int left = helper(root->left, parent, m);
            int right = helper(root->right, parent, m);
            
            m = max(m, left + right);
            return max(left + 1, right + 1);
        } else {
            int left = helper(root->left, root, m);
            int right = helper(root->right, root, m);
            m = max(m, left + right);
            return 0;
        }
    }
};
